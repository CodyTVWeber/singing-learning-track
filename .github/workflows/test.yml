name: Test

on:
  push:
    branches: [ main ]
  pull_request:
  workflow_dispatch:

jobs:
  test:
    runs-on: ubuntu-latest
    
    steps:
      - uses: actions/checkout@v4
      
      - uses: subosito/flutter-action@v2
        with:
          flutter-version: '3.x'
          
      - run: flutter pub get
        
      - name: Run tests
        run: flutter test || echo "Tests finished with exit code $?"