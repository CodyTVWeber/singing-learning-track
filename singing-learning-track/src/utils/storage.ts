import { User, Progress } from '../types';

const STORAGE_KEYS = {
  USER: 'singtrack_user',
  PROGRESS: 'singtrack_progress',
  SETTINGS: 'singtrack_settings',
};

// User Management
export const saveUser = (user: User): void => {
  localStorage.setItem(STORAGE_KEYS.USER, JSON.stringify(user));
};

export const getUser = (): User | null => {
  const userStr = localStorage.getItem(STORAGE_KEYS.USER);
  if (!userStr) return null;
  try {
    return JSON.parse(userStr);
  } catch {
    return null;
  }
};

export const createNewUser = (name: string, ageGroup: User['ageGroup']): User => {
  const user: User = {
    id: `user_${Date.now()}`,
    name,
    ageGroup,
    currentLevel: 1,
    totalPoints: 0,
  };
  saveUser(user);
  return user;
};

// Progress Management
export const saveProgress = (progress: Progress): void => {
  const allProgress = getAllProgress();
  const index = allProgress.findIndex(
    p => p.userId === progress.userId && p.lessonId === progress.lessonId
  );
  
  if (index >= 0) {
    allProgress[index] = progress;
  } else {
    allProgress.push(progress);
  }
  
  localStorage.setItem(STORAGE_KEYS.PROGRESS, JSON.stringify(allProgress));
};

export const getAllProgress = (): Progress[] => {
  const progressStr = localStorage.getItem(STORAGE_KEYS.PROGRESS);
  if (!progressStr) return [];
  try {
    return JSON.parse(progressStr);
  } catch {
    return [];
  }
};

export const getUserProgress = (userId: string): Progress[] => {
  return getAllProgress().filter(p => p.userId === userId);
};

export const getLessonProgress = (userId: string, lessonId: string): Progress | null => {
  return getAllProgress().find(
    p => p.userId === userId && p.lessonId === lessonId
  ) || null;
};

export const isLessonCompleted = (userId: string, lessonId: string): boolean => {
  const progress = getLessonProgress(userId, lessonId);
  return progress?.completed || false;
};

// User Stats
export const getUserStats = (userId: string) => {
  const progress = getUserProgress(userId);
  const completedLessons = progress.filter(p => p.completed).length;
  const totalScore = progress.reduce((sum, p) => sum + p.score, 0);
  const lastActivity = progress.length > 0 
    ? new Date(Math.max(...progress.map(p => new Date(p.completedDate).getTime())))
    : null;
  
  return {
    completedLessons,
    totalScore,
    lastActivity,
    streak: calculateStreak(userId),
  };
};

// Calculate daily streak
const calculateStreak = (userId: string): number => {
  const progress = getUserProgress(userId);
  if (progress.length === 0) return 0;
  
  // Sort by date descending
  const sortedDates = progress
    .map(p => new Date(p.completedDate).toDateString())
    .filter((date, index, self) => self.indexOf(date) === index)
    .sort((a, b) => new Date(b).getTime() - new Date(a).getTime());
  
  let streak = 0;
  const today = new Date().toDateString();
  const yesterday = new Date(Date.now() - 86400000).toDateString();
  
  // Check if user has activity today or yesterday
  if (sortedDates[0] !== today && sortedDates[0] !== yesterday) {
    return 0;
  }
  
  // Count consecutive days
  let currentDate = new Date(sortedDates[0]);
  for (const dateStr of sortedDates) {
    const date = new Date(dateStr);
    const dayDiff = Math.floor((currentDate.getTime() - date.getTime()) / 86400000);
    
    if (dayDiff <= 1) {
      streak++;
      currentDate = date;
    } else {
      break;
    }
  }
  
  return streak;
};

// Update user points
export const updateUserPoints = (userId: string, points: number): void => {
  const user = getUser();
  if (user && user.id === userId) {
    user.totalPoints += points;
    saveUser(user);
  }
};

// Clear all data (for testing/reset)
export const clearAllData = (): void => {
  localStorage.removeItem(STORAGE_KEYS.USER);
  localStorage.removeItem(STORAGE_KEYS.PROGRESS);
  localStorage.removeItem(STORAGE_KEYS.SETTINGS);
};