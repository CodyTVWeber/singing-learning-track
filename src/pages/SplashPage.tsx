import React, { useEffect, useState } from 'react';
import { useNavigate } from 'react-router-dom';
import { useApp } from '../context/AppContext';
import { LoadingSpinner } from '../components/LoadingSpinner';
import { colors, fontSize, fontWeight, spacing, gradients, shadows, animations } from '../theme/theme';
import { analytics } from '../services/analytics';
import { KookaImage } from '../components/KookaImage';

export const SplashPage: React.FC = () => {
  const navigate = useNavigate();
  const { user, isLoading } = useApp();
  const [showContent, setShowContent] = useState(false);
  
  // Kooka image handled by KookaImage component (randomized internally by default)

  useEffect(() => {
    // Track app open on initial mount
    analytics.trackEvent('app_opened');
    
    // Trigger content animation
    setTimeout(() => setShowContent(true), 100);

    if (!isLoading) {
      const timer = setTimeout(() => {
        if (user) {
          navigate('/skill-tree');
        } else {
          navigate('/onboarding');
        }
      }, 2500);

      return () => clearTimeout(timer);
    }
  }, [user, isLoading, navigate]);

  return (
    <div
      style={{
        display: 'flex',
        flexDirection: 'column',
        alignItems: 'center',
        justifyContent: 'center',
        minHeight: '100vh',
        background: gradients.soft,
        padding: spacing.xl,
        position: 'relative',
        overflow: 'hidden',
      }}
    >
      {/* Background decoration */}
      <div
        style={{
          position: 'absolute',
          top: '-20%',
          right: '-10%',
          width: '400px',
          height: '400px',
          background: gradients.wing,
          borderRadius: '50%',
          filter: 'blur(100px)',
          opacity: 0.3,
          animation: animations.pulse,
          pointerEvents: 'none',
        }}
      />
      <div
        style={{
          position: 'absolute',
          bottom: '-20%',
          left: '-10%',
          width: '400px',
          height: '400px',
          background: gradients.feather,
          borderRadius: '50%',
          filter: 'blur(100px)',
          opacity: 0.3,
          animation: animations.pulse,
          animationDelay: '1s',
          pointerEvents: 'none',
        }}
      />
      
      <div
        style={{
          opacity: showContent ? 1 : 0,
          transform: showContent ? 'scale(1)' : 'scale(0.8)',
          transition: 'all 0.6s cubic-bezier(0.68, -0.55, 0.265, 1.55)',
          textAlign: 'center',
          zIndex: 1,
        }}
      >
        <KookaImage
          alt="Kooka the Kookaburra waving hello"
          size={280}
          hover={false}
          style={{ marginBottom: spacing.xxxl, transform: 'rotate(-5deg)' }}
        />
        
        <h1
          style={{
            fontSize: fontSize.huge,
            fontWeight: fontWeight.extrabold,
            background: gradients.primary,
            WebkitBackgroundClip: 'text',
            WebkitTextFillColor: 'transparent',
            backgroundClip: 'text',
            marginBottom: spacing.md,
            textAlign: 'center',
            letterSpacing: '-0.03em',
            textShadow: '0 2px 10px rgba(0,0,0,0.1)',
          }}
        >
          Kooka Sing
        </h1>
        
        <p
          style={{
            fontSize: fontSize.xl,
            color: colors.textLight,
            marginBottom: spacing.xl,
            textAlign: 'center',
            maxWidth: '400px',
            lineHeight: 1.6,
            fontWeight: fontWeight.medium,
          }}
        >
          Join Kooka the Kookaburra on a magical journey to discover your singing voice!
        </p>
        
        <div style={{ marginTop: spacing.xl }}>
          <LoadingSpinner size="medium" variant="musical" />
          <p
            style={{
              fontSize: fontSize.sm,
              color: colors.textMuted,
              marginTop: spacing.md,
              fontStyle: 'italic',
            }}
          >
            Preparing your adventure...
          </p>
        </div>
      </div>

      {/* Musical notes floating animation */}
      <div
        style={{
          position: 'absolute',
          bottom: '10%',
          left: '10%',
          fontSize: '30px',
          animation: 'float 4s ease-in-out infinite',
          animationDelay: '0.5s',
          opacity: 0.3,
          pointerEvents: 'none',
        }}
      >
        <img
          src="/img/kooka-burra-breathing.png"
          alt="Floating Kooka"
          style={{
            width: '40px',
            borderRadius: '50%',
            transform: 'rotate(15deg)',
          }}
        />
      </div>
      <div
        style={{
          position: 'absolute',
          top: '20%',
          right: '15%',
          fontSize: '25px',
          animation: 'float 5s ease-in-out infinite',
          animationDelay: '1.5s',
          opacity: 0.15,
          pointerEvents: 'none',
        }}
      >
        <img
          src="/img/kooka-burra-calling-out.png"
          alt="Floating Kooka"
          style={{
            width: '50px',
            borderRadius: '50%',
            transform: 'rotate(-10deg)',
          }}
        />
      </div>
      <div
        style={{
          position: 'absolute',
          top: '50%',
          left: '5%',
          fontSize: '20px',
          animation: 'float 6s ease-in-out infinite',
          animationDelay: '2s',
          opacity: 0.1,
          pointerEvents: 'none',
        }}
      >
        <img
          src="/img/kooka-burra-waiving.png"
          alt="Floating Kooka"
          style={{
            width: '35px',
            borderRadius: '50%',
            transform: 'rotate(25deg)',
          }}
        />
      </div>
    </div>
  );
};