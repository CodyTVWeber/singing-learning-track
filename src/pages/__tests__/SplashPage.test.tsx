import React from 'react';
import { renderWithProviders } from '../../test/utils';
import { SplashPage } from '../SplashPage';

describe('SplashPage', () => {
  it('renders and matches snapshot', () => {
    const { asFragment } = renderWithProviders(<SplashPage />);
    expect(asFragment()).toMatchSnapshot();
  });
});

