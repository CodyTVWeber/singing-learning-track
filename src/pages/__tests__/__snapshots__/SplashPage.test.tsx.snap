// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SplashPage > renders and matches snapshot 1`] = `
<DocumentFragment>
  <div
    style="display: flex; flex-direction: column; align-items: center; justify-content: center; min-height: 100vh; background: linear-gradient(135deg, #FAF6F2 0%, #F2E8DD 100%); padding: 32px; position: relative; overflow: hidden;"
  >
    <div
      style="position: absolute; top: -20%; right: -10%; width: 400px; height: 400px; background: linear-gradient(135deg, #4B708B 0%, #6B8CAB 100%); border-radius: 50%; filter: blur(100px); opacity: 0.3; animation: pulse 2s ease-in-out infinite; pointer-events: none;"
    />
    <div
      style="position: absolute; bottom: -20%; left: -10%; width: 400px; height: 400px; background: linear-gradient(135deg, #6B5843 0%, #8B7355 50%, #D4C4B0 100%); border-radius: 50%; filter: blur(100px); opacity: 0.3; animation: pulse 2s ease-in-out infinite; animation-delay: 1s; pointer-events: none;"
    />
    <div
      style="opacity: 0; transform: scale(0.8); transition: all 0.6s cubic-bezier(0.68, -0.55, 0.265, 1.55); text-align: center; z-index: 1;"
    >
      <img
        alt="Kooka the Kookaburra waving hello"
        src="/img/kooka-burra-waiving.png"
        style="width: 280px; filter: drop-shadow(0 20px 25px -5px rgba(0, 0, 0, 0.08), 0 10px 10px -5px rgba(0, 0, 0, 0.02)); transform: rotate(-5deg); margin-bottom: 64px;"
      />
      <h1
        style="font-size: 64px; font-weight: 800; background: linear-gradient(135deg, #8B7355 0%, #A68B68 100%); -webkit-text-fill-color: transparent; background-clip: text; margin-bottom: 16px; text-align: center; letter-spacing: -0.03em; text-shadow: 0 2px 10px rgba(0,0,0,0.1);"
      >
        Kooka Sing
      </h1>
      <p
        style="font-size: 24px; color: rgb(107, 88, 67); margin-bottom: 32px; text-align: center; max-width: 400px; line-height: 1.6; font-weight: 500;"
      >
        Join Kooka the Kookaburra on a magical journey to discover your singing voice!
      </p>
      <div
        style="margin-top: 32px;"
      >
        <div
          style="display: inline-flex; gap: 5px; align-items: flex-end; height: 60px;"
        >
          <div
            style="width: 7.5px; height: 100%; background: linear-gradient(135deg, #4B708B 0%, #6B8CAB 100%); border-radius: 3.75px; animation: pulse 1.2s ease-in-out 0s infinite; transform-origin: bottom;"
          />
          <div
            style="width: 7.5px; height: 100%; background: linear-gradient(135deg, #4B708B 0%, #6B8CAB 100%); border-radius: 3.75px; animation: pulse 1.2s ease-in-out 0.15s infinite; transform-origin: bottom;"
          />
          <div
            style="width: 7.5px; height: 100%; background: linear-gradient(135deg, #4B708B 0%, #6B8CAB 100%); border-radius: 3.75px; animation: pulse 1.2s ease-in-out 0.3s infinite; transform-origin: bottom;"
          />
          <div
            style="width: 7.5px; height: 100%; background: linear-gradient(135deg, #4B708B 0%, #6B8CAB 100%); border-radius: 3.75px; animation: pulse 1.2s ease-in-out 0.44999999999999996s infinite; transform-origin: bottom;"
          />
        </div>
        <p
          style="font-size: 14px; color: rgb(139, 115, 85); margin-top: 16px; font-style: italic;"
        >
          Preparing your adventure...
        </p>
      </div>
    </div>
    <div
      style="position: absolute; bottom: 10%; left: 10%; font-size: 30px; animation: float 4s ease-in-out infinite; animation-delay: 0.5s; opacity: 0.3; pointer-events: none;"
    >
      <img
        alt="Floating Kooka"
        src="/img/kooka-burra-breathing.png"
        style="width: 40px; border-radius: 50%; transform: rotate(15deg);"
      />
    </div>
    <div
      style="position: absolute; top: 20%; right: 15%; font-size: 25px; animation: float 5s ease-in-out infinite; animation-delay: 1.5s; opacity: 0.15; pointer-events: none;"
    >
      <img
        alt="Floating Kooka"
        src="/img/kooka-burra-calling-out.png"
        style="width: 50px; border-radius: 50%; transform: rotate(-10deg);"
      />
    </div>
    <div
      style="position: absolute; top: 50%; left: 5%; font-size: 20px; animation: float 6s ease-in-out infinite; animation-delay: 2s; opacity: 0.1; pointer-events: none;"
    >
      <img
        alt="Floating Kooka"
        src="/img/kooka-burra-waiving.png"
        style="width: 35px; border-radius: 50%; transform: rotate(25deg);"
      />
    </div>
  </div>
</DocumentFragment>
`;
